src/pubsub/ua_pubsub.c

```diff
@@ -182,6 +185,1098 @@ UA_Server_removeWriterGroup(UA_Server *server, const UA_NodeId writerGroup){
+ * @param readerGroupIdentifier
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode
+UA_PubSubConnection_addReaderGroup(UA_Server *server, UA_NodeId connectionIdentifier,
+                                   const UA_PubSubReaderGroupConfig *readerGroupConfig,
+                                   UA_NodeId *readerGroupIdentifier) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_PubSubReaderGroupConfig tmpReaderGroupConfig;
+
+    //Search the connection by the given connectionIdentifier
+    if (!readerGroupConfig) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    //Search the connection by the given connectionIdentifier
+    UA_PubSubConnection *currentConnectionContext = UA_PubSubConnection_findConnectionbyId(server, connectionIdentifier);
+    if (!currentConnectionContext) {
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+
+    //Allocate memory for new reader group
+    UA_PubSubReaderGroup *newGroup = (UA_PubSubReaderGroup *)UA_calloc(1, sizeof(UA_PubSubReaderGroup));
+    if (!newGroup) {
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+    }
+
+    //Generate nodeid for the readergroup identifier
+    newGroup->linkedConnection = currentConnectionContext->identifier;
+    UA_PubSubManager_generateUniqueNodeId(server, &newGroup->identifier);
+    if (readerGroupIdentifier) {
+        UA_NodeId_copy(&newGroup->identifier, readerGroupIdentifier);
+    }
+
+    //Deep copy of the config
+    retval |= UA_PubSubReaderGroupConfig_copy(readerGroupConfig, &tmpReaderGroupConfig);
+    newGroup->config = tmpReaderGroupConfig;
+    retval |= UA_ReaderGroup_addSubscribeCallback(server, newGroup);
+    LIST_INSERT_HEAD(&currentConnectionContext->readerGroups, newGroup, listEntry);
+    currentConnectionContext->readerGroupsSize++;
+
+#ifdef UA_ENABLE_PUBSUB_INFORMATIONMODEL
+    addReaderGroupRepresentation(server, newGroup);
+#endif
+
+    return retval;
+}
+
+/**
+ * Remove ReaderGroup from connection and delete contained readers.
+ *
+ * @param server
+ * @param groupIdentifier
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode
+UA_PubSubConnection_removeReaderGroup(UA_Server *server, UA_NodeId groupIdentifier) {
+    UA_PubSubReaderGroup* readerGroup = UA_PubSubReaderGroup_findRGbyId(server, groupIdentifier);
+    if (readerGroup == NULL) {
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+
+    //Search the connection to which the given readergroup is connected to
+    UA_PubSubConnection *connection = UA_PubSubConnection_findConnectionbyId(server, readerGroup->linkedConnection);
+    if (connection == NULL) {
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+
+#ifdef UA_ENABLE_PUBSUB_INFORMATIONMODEL
+    //ToDo:RemoveGroupRepresentation(server, &readerGroup->identifier);
+#endif
+
+    //UA_PubSubReaderGroup_delete also removes itself from the list
+    UA_PubSubReaderGroup_delete(server, readerGroup);
+    //Remove readerGroup from Connection
+    LIST_REMOVE (readerGroup, listEntry);
```

Comment:
Avoid spaces before braces to be consistent with the remaining library.
At least for function calls and macros (and also before if/for/...).